ext/opentelemetry-ext-opentracing-shim/src/opentelemetry/ext/opentracingshim/__init__.py

```diff
@@ -0,0 +1,257 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from contextlib import contextmanager
+
+import opentracing
+
+from opentelemetry.ext.opentracingshim import util
+
+logger = logging.getLogger(__name__)
+
+
+def create_tracer(tracer, scope_manager=None):
+    return TracerWrapper(tracer, scope_manager)
+
+
+class SpanContextWrapper(opentracing.SpanContext):
+    def __init__(self, otel_context):
+        self._otel_context = otel_context
+
+    def unwrap(self):
+        """Returns the wrapped OpenTelemetry `SpanContext` object."""
+
+        return self._otel_context
+
+    @property
+    def baggage(self):
+        logger.warning(
+            "Using unimplemented property baggage on class %s.",
+            self.__class__.__name__,
+        )
+        # TODO: Implement.
+
+
+class SpanWrapper(opentracing.Span):
+    def __init__(self, tracer, context, span):
+        super().__init__(tracer, context)
+        self._otel_span = span
+
+    def unwrap(self):
+        """Returns the wrapped OpenTelemetry `Span` object."""
+
+        return self._otel_span
+
+    def set_operation_name(self, operation_name):
+        self._otel_span.update_name(operation_name)
+
+        # Return self for call chaining.
+        return self
+
+    def finish(self, finish_time=None):
+        end_time = finish_time
+        if end_time is not None:
+            end_time = util.time_seconds_to_ns(finish_time)
+        self._otel_span.end(end_time=end_time)
+
+    def set_tag(self, key, value):
+        self._otel_span.set_attribute(key, value)
+
+        # Return self for call chaining.
+        return self
+
+    def log_kv(self, key_values, timestamp=None):
+        if timestamp is not None:
+            event_timestamp = util.time_seconds_to_ns(timestamp)
+        else:
+            event_timestamp = None
+
+        event_name = util.event_name_from_kv(key_values)
+        self._otel_span.add_event(event_name, event_timestamp, key_values)
+
+        # Return self for call chaining.
+        return self
+
+    def set_baggage_item(self, key, value):
+        logger.warning(
+            "Calling unimplemented method set_baggage_item() on class %s",
+            self.__class__.__name__,
+        )
+        # TODO: Implement.
+
+    def get_baggage_item(self, key):
+        logger.warning(
+            "Calling unimplemented method get_baggage_item() on class %s",
+            self.__class__.__name__,
+        )
+        # TODO: Implement.
+
+    # TODO: Verify calls to deprecated methods `log_event()` and `log()` on
+    # base class work properly (it's probably fine because both methods call
+    # `log_kv()`).
+
+
+class ScopeWrapper(opentracing.Scope):
+    def close(self):
+        self._span.finish()
```

Comment:
As mentioned somewhere else, we need to handle the case where you don't actually finish a Span as part of closing the Scope ;)

Reply:
I think this should be handled as of https://github.com/kinvolk/opentelemetry-python/commit/fea80ea5c553e6d88280524e6a30e28610785096.

We now override the `__exit__()` method of `opentracing.Scope`, so `close()` is no longer called by `__exit__()`. Rather, `__exit__()` on `opentracing.Scope` calls `__exit__()` on the OTel context manager returned by `use_span()`, which contains a [check](https://github.com/open-telemetry/opentelemetry-python/blob/b0da53d28b041338e04982f3a9715926887b5cad/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L385-L386) for `end_on_exit`. `close()` is now used only when the instrumentation explicitly calls `ScopeWrapper.close()`.

Please tell me if this looks right to you @carlosalberto.

Reply:
Hey, commenting again as I saw my comments hadn't been erased ;)

Well, you are being too optimistic. OT users can still manually do `Scope.close()`, so please add a note there.

Reply:
>Well, you are being too optimistic. OT users can still manually do Scope.close(), so please add a note there.

If a user calls `close()` on a `ScopeShim` object (renamed from `ScopeWrapper`), the span will end as expected, unless I am missing your point. This is verified by the tests:

https://github.com/kinvolk/opentelemetry-python/blob/6dac24d1b16ee419bb063fbcf8e42afd903feb29/ext/opentelemetry-ext-opentracing-shim/tests/test_shim.py#L207-L219

What note should I add?

Reply:
Actually, I think we should not override `__exit__` at all, but implement the logic in `close`. It also seems that we must then call the wrapped scope's `__enter__` already in `__init__` and remove our override of `__enter__`.

Reply:
@Oberon00 I've refactored `ScopeShim` according to your suggestion. Please tell me if this is what you meant.
One downside of this approach though is that since we no longer override `__exit__` on `opentracing.Scope` we can't pass the error information (`exc_type`, `exc_val`, `exc_tb`) on to `__exit__` on the OTel span context manager and we pass `None` in all arguments instead. For the tracer implementation in the OTel SDK it's OK since it doesn't use these arguments, but in case another implementation needs these values, looks like we won't have a choice but to override `__exit__` on `opentracing.Scope`.

@carlosalberto following my latest refactor, are your concerns regarding not finishing a span addressed now?

Reply:
Yes, that is what I meant! 👍 

True, the exception info thing is a bit of a deficiency. 🤔 We could work around this by going back to overriding exit, saving the arguments in the ScopeShim object and then calling `super().__exit__`. Then we could pick up the saved arguments info in `close`, use them and clear them in a `finally` in `__exit__`. That sounds like it should work well enough but also like a hack. 😅 Since we don't currently use the exception info in `__exit__`, I'd say it's your call whether to change this in this PR or not.

Reply:
My preference would be to do this change if and when it becomes necessary. Thanks!